File: topos/mcp/refactor_hotspots.py
"""Shared markdown renderer for the ``topos_refactor_*`` tool family.

Mirrors how ``render_evaluation_md`` (``formatting.py``) is the single
renderer reused by every ``topos_evaluate_*`` tool: all three refactor tools
(cycles, dependencies, process) report the same "ranked hotspot" row shape,
so they share one table renderer instead of three near-duplicates.
"""

from __future__ import annotations

from .schemas import RefactorHotspot


def render_hotspots_md(title: str, hotspots: list[RefactorHotspot]) -> str:
    """Render a ranked list of :class:`RefactorHotspot` rows as markdown."""
    if not hotspots:
        return f"**{title}:** none found."

    lines = [f"## {title}", "", "| Kind | Label | Location | Score | Suggestion |"]
    lines.append("| --- | --- | --- | ---: | --- |")
    for h in hotspots:
        location = h.filepath
        if h.line_start is not None:
            location += f":{h.line_start}"
            if h.line_end is not None and h.line_end != h.line_start:
                location += f"-{h.line_end}"
        safe_label = h.label.replace("\n", " ").replace("|", "\\|")
        safe_suggestion = h.suggestion.replace("\n", " ").replace("|", "\\|")
        row = (
            f"| `{h.kind}` | `{safe_label}` | `{location}` | "
            f"{h.score:.3f} | {safe_suggestion} |"
        )
        lines.append(row)
    return "\n".join(lines)

File: topos/mcp/src/refactor_hotspots.rs
//! Shared markdown renderer for the `topos_refactor` tool family.
//!
//! All three refactor targets (cycles, dependencies, process) report the
//! same "ranked hotspot" row shape, so they share one table renderer.

use crate::schemas::RefactorHotspot;

/// Render a ranked list of [`RefactorHotspot`] rows as markdown.
pub fn render_hotspots_md(title: &str, hotspots: &[RefactorHotspot]) -> String {
    if hotspots.is_empty() {
        return format!("**{title}:** none found.");
    }

    let mut lines = vec![
        format!("## {title}"),
        String::new(),
        "| Kind | Label | Location | Score | Suggestion |".to_string(),
        "| --- | --- | --- | ---: | --- |".to_string(),
    ];
    for h in hotspots {
        let mut location = h.filepath.clone();
        if let Some(start) = h.line_start {
            location.push_str(&format!(":{start}"));
            if let Some(end) = h.line_end {
                if end != start {
                    location.push_str(&format!("-{end}"));
                }
            }
        }
        let safe_label = h.label.replace('\n', " ").replace('|', "\\|");
        let safe_suggestion = h.suggestion.replace('\n', " ").replace('|', "\\|");
        lines.push(format!(
            "| `{}` | `{safe_label}` | `{location}` | {:.3} | {safe_suggestion} |",
            h.kind, h.score
        ));
    }
    lines.join("\n")
}
